proposals/exception-handling/Exceptions-formal-examples.md

```diff
@@ -0,0 +1,369 @@
+
+## Example 0.
+
+The only example with an almost full reduction trace, and all new instructions. Such explicit reduction steps are only shown in Example 4 and Example 5, to highlight the reduction step of the administrative `delegate`.
+
+In the following reduction, we don't show the first 4 steps, which just reduce the several `try`s and the `throw x` to their respective administrative instructions.
+
+```
+(func (result i32) (local i32)
+  try (result i32)
+    try
+      try
+        throw x
+      catch_all
+        i32.const 27
+        local.set 0
+        rethrow 0
+      end
+    delegate 0
+  catch x
+    local.get 0
+  end)
+```
+
+We write the body of this function in folded form, because it is easier to parse.
+
+```
+(try (result i32)
+  (do
+    (try
+      (do
+        (try
+          (do
+            (throw x))
+          (catch_all
+            (local.set 0 (i32.const 27))
+            (rethrow 0))))
+      (delegate 0)))
+  (catch x
+    (local.get 0)))
+```
+
+Take the frame `F = (locals i32.const 0, module m)`. We have:
+
+```
+↪ ↪ ↪
+↪ F; (label_1{}
+        (catch_1{ a_x (local.get 0) }
+          (label_0{}
+            (delegate{ 0 }
+              (label_0{}
+                (catch_0{ ε (local.set 0 (i32.const 27)) (rethrow 0) }
+                  (throw a_x) end) end) end) end) end) end)
+```
+
+For the trivial throw context `T = [_]` the above is the same as
+
+```
+↪ F; (label_1{}
+        (catch_1{ a_x (local.get 0) }
+          (label_0{}
+            (delegate{ 0 }
+              (label_0{}
+                (catch_0{ ε (local.set 0 (i32.const 27)) (rethrow 0) }
+                  T[(throw a_x)]) end) end) end) end) end)
+
+↪ F; (label_1{}
+        (catch_1{ a_x (local.get 0) }
+          (label_0{}
+            (delegate{ 0 }
+              (label_0{}
+                (caught_0{ a_x ε }
+                  (local.set 0 (i32.const 27))
+                  (rethrow 0) end) end) end) end) end) end)
+```
+
+Let `F'` be the frame `{locals i32.const 27, module m}`, and let `B^0 = [_]` to reduce `rethrow 0`.
+
+```
+↪ F'; (label_1{}
+         (catch_1{ a_x (local.get 0) }
+           (label_0{}
+             (delegate{ 0 }
+               (label_0{}
+                 (caught_0{ a_x ε }
+                   B^0[ (rethrow 0) ] end) end) end) end) end) end)
+
+↪ F'; (label_1{}
+         (catch_1{ a_x (local.get 0) }
+           (label_0{}
+             (delegate{ 0 }
+               (label_0{}
+                 (caught_0{ a_x }
+                   (throw a_x) end) end) end) end) end) end)
+```
+
+Let `T' = (label_0{} (caught_0{ a_x ε } [_] end) end)` and use the same `B^0` as above to reduce the throw with the delegate.
+
+```
+↪ F'; (label_1{}
+         (catch_1{ a_x (local.get 0) }
+           (label_0{}
+             B^0[ (delegate{ 0 } T'[ (throw a_x) ] end) ] end) end) end)
```

Comment:
Hmm, I think I understand. Thanks. And I don't think you should worry about "single comments"; I believe most email clients should group mails into threads anyway; otherwise they would be unusable. Gmail web client does that too. I also do "single comments" all the time...